gips/0081-indexing-payments.md

```diff
@@ -0,0 +1,165 @@
+
+In this MVP, we will roll out a new gateway component (the “dipper”) that will be controlled with a CLI. Gateway operators can be onboarded to this CLI and use it to add subgraphs to the list of supported subgraphs. Once a subgraph is in the list, the dipper will take care of finding Indexers and setting up indexing agreements with them.
+
+The Indexer stack will be updated to be able to accept indexing agreements from the dipper when the price is above the price set by the Indexer, and up to a maximum amount of subgraphs also set by the Indexer.
+
+Pricing between Indexer and gateway in the MVP will be based on the "amount of work done" proxy detailed above.
+
+The dipper will initially set up agreements with 3 Indexers for each desired subgraph, as we estimate this is a reasonable number to achieve 99.9% uptime in most subgraphs.
+
+Once the smart contracts proposed in this GIP are deployed, we anticipate adapting the indexer and gateway components accordingly.
+
+The MVP has different trust implications than what this proposal has described so far. In particular these are:
+
+- Indexers do the work upfront and trust that the gateway will pay them accordingly after the fact, since the gateway issues payment vouchers upon receiving PoIs and not before.
+- Gateways trust that indexers are reporting the correct amount of work done, since this information is not posted on-chain and thus it's non-slashable. This is partially mitigated by the fact that the gateway can (theoretically) compare the reported work done among the pool of indexers selected and thus can weed out bad actors, but it's nevertheless a weaker mechanism than slashing.
+
+## Detailed Specification
+
+TODO
+
+## Dependencies
+
+This GIP depends on Horizon being fully functional.
+
+## Rationale and Alternatives
+
+When presenting this proposal, we often hear suggestions for alternative, and possibly simpler, approaches to payments, for instance:
+
+- Could the Gateway just pre-pay a fixed amount per subgraph?
+- Could Indexers just charge a specific (Indexer-determined) amount for each subgraph in arrears?
+
+However, the design of The Graph protocol is, and must be, adversarial. When designing incentives, participants are generally considered rational, but the protocol must be safe against byzantine participants that may derive profit from malicious actions in ways that cannot be predicted within the protocol. The typical cryptoeconomic approach for this is to establish a cost-of-corruption, so that malicious actors are punished and participants can quantify the risk in each interaction.
```

Comment:
I would be tempted to say trust-minimized. The protocol is designed to be fault-resistent. While disputes and dispute resolution typically have adversariel mechanisms, that is different from designing the protocol to be adversariel. The protocol is, IMO, designed incentivize co-operative behaviour, while protecting against malicious behaviour. My protocol design is not designed to be adversariel, and the distinction is important.

The protocol being safe against byzantine participants is not a 'but' to considering participants rational. If anything, it is a 'because'. Not needed here though, so I suggest just removing.

Byzanine is a type of fault tolerance (BFT), rather than being a type of participant. It is not a type of fault-tolerance being implemented here, it is about consensus mechanisms rather than dispute resolution.
I've suggested a simplification. It's not perfect, but I think closer to being correct, and likely enough for your point?
```suggestion
However, the design of The Graph protocol is, and must be, trust-minimized. When designing incentives the protocol must protect against malicious actors whom might otherwise profit in ways that cannot be predicted within the protocol. The typical cryptoeconomic approach for this is to establish a cost-of-corruption, so that malicious actors are punished and participants can quantify the risk in each interaction.
```
```suggestion
However, the design of The Graph protocol is, and must be, adversarial. When designing incentives, participants are generally considered rational, but the protocol must be safe against byzantine participants that may derive profit from malicious actions in ways that cannot be predicted within the protocol. The typical cryptoeconomic approach for this is to establish a cost-of-corruption, so that malicious actors are punished and participants can quantify the risk in each interaction.
```

Reply:
@pcarranzav

Reply:
> While disputes and dispute resolution typically have adversariel mechanisms, that is different from designing the protocol to be adversariel. The protocol is, IMO, designed incentivize co-operative behaviour, while protecting against malicious behaviour. My protocol design is not designed to be adversariel, and the distinction is important.

I'm not sure I understand the difference. The protocol design is adversarial in that it considers that there are adversaries actively trying to attack it and must protect itself from that. That has always been my (intuitive) understanding of adversarial design. ChatGPT described it this way too:

> Designing software in an adversarial way typically refers to creating it with the assumption that its environment, users, or even developers might act maliciously or unpredictably. This design philosophy aims to ensure the system is robust against attacks, errors, or intentional misuse. It is common in areas like security-critical systems, cryptographic protocols, and blockchain applications.

Of course cooperative behavior is super important too, but that doesn't remove the need for adversarial thinking / design.

I'm also not sure I understand the note on Byzantine participants. Byzantine _is_ a type of participant or used that way in the literature, see https://arxiv.org/pdf/1910.08820 - and this paragraph is a nod to that paper actually, that illustrates my thinking behind this section. Considering Byzantine participants explicitly is important because if you're modeling only with rational participants, you might be tempted to remove slashing or not minimize trust: rational actors just want to profit so will behave according to the rules if it gives them more income. Maybe linking to that paper would help?

Reply:
On adversarial design, it is not such a big deal, but there is a subtle difference (at least in implication) between 'adversarial design' and 'the design being adversarial'. I would be more comfortable with saying I **do** adversarial design than that I **have** an adversarial design, although there is an area of overlap in meaning where you could consider them both to relate to the same activity. I do the first, that is a good thing. I avoid the second, it would be a bad thing. But fuzzy terminology (or worse, thinking) can conflate the two. It sounds like you are only thinking of the first. Do you understand the second, and why that would be bad (or at least represent a different approach, and not one I suggest)?

Regarding the article, I've seen it before. I'm not that impressed TBH. First I'll note it a has a click-bait title that the article does not show to be the case. The sloppy and equivocal misuse of terminology extends beyond the title to the content. The article seems confused about what a rational actor is, and wallows in this confusion and ongoing sloppy terminology use to ramble on about not much of substance. The main point being made is not invalid, but is not new insight, and presented in a verbose waffly way. While the authors state they do not provide formal proof but just an intuitve sketch of an argument, I think it doesn't even do that well. The article might be helpful to some, such flawed thinking in design is common. I think it is poorly constructed waffle, made to appear more profound than it really is by equivocal terminology misuse, and does not contribute helpful new thinking. I do not think it presents a credible source to justify use of terminology, but rather a cationary tale of the mess bad use of terminology gets you into.

Reply:
Thanks for clarifying on adversarial design. I could reword to say "the protocol must be designed in an adversarial way"; would that address your concern?

re: the article, clearly we got very different impressions when reading it 😅 
I can see your point about how it's written, but I personally learned/understood three useful things from that article:
1) The risk / attack vector of someone shorting a token as a general way to derive profit from any malicious activity
2) As a consequence of this (and other potential incentives outside the system), the risk of modeling all protocol participants as rational when it comes to security, and
3) The idea of modeling participants as Byzantine-Altruistic-Rational (this actually comes from this article cited there: https://citeseerx.ist.psu.edu/document?repid=rep1&type=pdf&doi=2865c128e0bdda6dabd043ff17a485f29d6d852d - which also talks about "Byzantine users" as a category).

This makes the terminology useful and the conclusions are useful to keep in mind when designing the protocol, even if they may seem obvious in retrospect.

Whether we cite the article or not, I think these are useful things to mention because I've often seen people making arguments that assume participants are rational and would open up risks from "Byzantine behavior".

Reply:
Regarding adversarial, I would say with an adversarial mindset. But really it is up to you, I share my opinion, but you do not have to word it how I would.

I agree the first paper highlights valid points, but I am not impressed by either its clarity or its logic. I've probobly gone enough into why, but I'll add one of the issues is the misguided and equivocal of what a 'rational actor'. For point 2 you refer to a risk of modeling all protocol participants as rational. This is based on the weird starting assumption that rationality is only about outcomes within the system. While that might well be a mistake that is sometimes made, I do not believe that is the usual starting assumption for a rational actor, quite the contrary. It should certainly not be part of the definition of 'rational actor'! Part of the problem of the paper is that it based on attacking a strawman.

I see the second paper does refer to Byzantine users, although mainly to Byzantine nodes, and the framework makes sense. (This paper is much better written than the other one.) That does (to me) represent valid precendent contrary to my original claim that Byzantine is not a type of participant. I caution however that BAR is about nodes in system subject to BFT analysis, a particular category of consensus mechanisms. We are not desribing such a consensus based system, the term is being used outside the context for which it made sense to categorise participants as BAR. For our context, the restricted assumptions for a rational node of that paper do not apply either. The BAR framework is not about the type of mechanism being discussed. While there is overlap in concepts, I am not sure it is wise to conflate the thinking. I worry that leads to a fuzzy lack of clarity and distinction that creates a risk of misapplying the conclusions of that paper.

Reply:
> This is based on the weird starting assumption that rationality is only about outcomes within the system. While that might well be a mistake that is sometimes made, I do not believe that is the usual starting assumption for a rational actor, quite the contrary.

I agree it should not be, especially when thinking about security. But when modeling the protocol as a game to understand the incentives, it is often a reasonable simplification to think only about the actions and rewards that are present in the system. This will at least let you know if you have a mechanism where the dominant strategy is not what you want participants to do...

> I caution however that BAR is about nodes in system subject to BFT analysis, a particular category of consensus mechanisms.

I agree, but I think the framework is still super useful, and we can (carefully) apply it to the things we do, e.g. we could look at the risks of hiring a Sybil for a subgraph, or the risk of getting incorrect data, based on some assumption of how many Indexers are Byzantine vs. rational.

Anyways, my intent is not to defend that paper. I like to think in terms of "Byzantine participants", and I think it's important to make the point that you need to add protections like slashing even if it looks like a rational actor would be incentivized to act honestly. But it's clear this paragraph needs some rewording.

Reply:
I reworded this section a bit, wdyt @RembrandtK?